### PR TITLE
Optimize CONCAT(CONCAT(a, b), c)

### PR DESCRIPTION
**Pull request by raff**
Optimize CONCAT(CONCAT(a, b), c) reported in https://github.com/benhoyt/goawk/issues/91

Added a ConcatN opcode. Could have used MultiIndex opcode (I tried and it works
as is) but "concat" should not be using a separator.

### CODE REVIEW

**Comment by benhoyt** [comment]
Nice, and thank you! I'll take a closer look at this in the next few days.

File: interp/vm.go [hunk 1]
```diff
@@ -414,6 +414,16 @@ func (p *interp) execute(code []compiler.Opcode) error {
 			l, r := p.peekPop()
 			p.replaceTop(str(p.toString(l) + p.toString(r)))
 
+		case compiler.ConcatN:
+			numValues := int(code[ip])
+			ip++
+			values := p.popSlice(numValues)
+			parts := make([]string, 0, numValues)
```
Comment (benhoyt) [inline interp/vm.go:421]: Probably more efficient to use `strings.Builder` and `WriteString` here. (Though maybe you can test that assumption!)

File: internal/compiler/opcodes.go [hunk 2]
```diff
@@ -68,6 +68,9 @@ const (
 	// Multi-index concatenation
 	MultiIndex // num
 
+	// Multi-value concatenation
+	ConcatN
```
Comment (benhoyt) [inline internal/compiler/opcodes.go:72]: Let's consider just making `Concat` always take an argument (numValues), assuming that doesn't slow down the concat-two case significantly. Could you please add a benchmark to interp_test.go `BenchmarkConcatTwo` that benchmarks this case, and post before/after results?

**Comment by benhoyt** [comment]
Or maybe I'll get to it while the iron's hot. :-) Added a couple of review comments. Could you also please add the benchmark I mentioned and then post your before/after results of running all three benchmarks (ideally with `-count=5` and showing the diff using the benchstat tool). Thanks!

File: internal/compiler/opcodes.go (hunk 2 above)
Comment (benhoyt) [reply internal/compiler/opcodes.go:72]: You'll also need to add ConcatN (or Concat if that's what we end up calling it) to the disassembler, so it consumes and displays the numValues argument.

File: internal/compiler/compiler.go [hunk 3]
```diff
@@ -875,6 +877,35 @@ func (c *compiler) expr(expr ast.Expr) {
...
+	for {
+		if be, ok := e.Left.(*ast.BinaryExpr); ok && be.Op == lexer.CONCAT {
+			parts = append(parts, e.Right)
+			e = be
+		} else {
+			parts = append(parts, e.Right)
+			parts = append(parts, e.Left)
+			break
+		}
+	}
```
Comment (benhoyt) [inline internal/compiler/compiler.go:893]: What about this structure (and variable naming)? IMO it's a bit clearer and avoids the `else` (the early-exit pattern is more common in Go).

```go
func (c *compiler) concatOp(expr *ast.BinaryExpr) {
	var values []ast.Expr
	for {
		values = append(values, expr.Right)
		left, isBinary := expr.Left.(*ast.BinaryExpr)
		if !isBinary || left.Op != lexer.CONCAT {
			break
		}
		expr = left
	}
	values = append(values, expr.Left)
        ...
```

File: internal/compiler/compiler.go [hunk 4]
```diff
@@ -875,6 +877,35 @@ func (c *compiler) expr(expr ast.Expr) {
...
+
+	if len(parts) == 2 {
+		c.expr(parts[1])
+		c.expr(parts[0])
+		c.add(Concat)
+		return
+	}
+
+	for i := len(parts) - 1; i >= 0; i-- {
```
Comment (benhoyt) [inline internal/compiler/compiler.go:902]: Might be good to add a comment here that `// we appended the values in reverse above`.

File: internal/compiler/opcodes.go (hunk 2 above)
Comment (raff) [reply internal/compiler/opcodes.go:72]: The original concat with 2 arguments seems to be at least twice as fast than using the new one:

`
// ConcatN
  BenchmarkConcatTwo-16               	 8443903	       139.6 ns/op
  BenchmarkConcatTwo-16               	 8600488	       137.8 ns/op
  BenchmarkConcatTwo-16               	 8743024	       141.4 ns/op
  BenchmarkConcatTwo-16               	 8841157	               138.4 ns/op
  BenchmarkConcatTwo-16               	 8648384	       139.0 ns/op
---
// Concat
 BenchmarkConcatTwo-16               	15598926	        74.76 ns/op
 BenchmarkConcatTwo-16               	16030410	        76.17 ns/op
 BenchmarkConcatTwo-16               	15828648	        74.84 ns/op
 BenchmarkConcatTwo-16               	16167350	        76.84 ns/op
 BenchmarkConcatTwo-16               	15776442	        75.88 ns/op
`

File: internal/compiler/compiler.go (hunk 3 above)
Comment (raff) [reply internal/compiler/compiler.go:893]: Yup! This looks nicer!

File: internal/compiler/opcodes.go (hunk 2 above)
Comment (raff) [reply internal/compiler/opcodes.go:72]: Dropping to 98 when using strings.Builder in the VM interpreter:

`
BenchmarkConcatTwo-16               	12588488	        97.84 ns/op
BenchmarkConcatTwo-16               	12410306	        98.79 ns/op
BenchmarkConcatTwo-16               	12267706	        98.20 ns/op
BenchmarkConcatTwo-16               	12534445	        98.70 ns/op
BenchmarkConcatTwo-16               	12488019	        98.12 ns/op
`

File: interp/vm.go (hunk 1 above)
Comment (raff) [reply interp/vm.go]: Done!

**Comment by raff** [comment]
Let me know if you want to go with a single Concat that supports 2+ parameters or keep 2 separate ones, given the benchmark results.

**Comment by benhoyt** [comment]
@raff Thanks for the updates. Yeah, I think having a separate ConcatN makes sense with those numbers. Maybe rename `Concat` to `Concat2`?

Could you also please post your before/after results for the other BenchmarkConcat benchmarks?

**Comment by raff** [comment]
Do you want Concat2 and ConcatN or Concat2 and Concat ?

I just figured out where to get benchstat, I'll post the before/after results in a bit.

Not sure of why when enabling/disabling the original concat I get a ~4% difference in ConcatSmall:

```
ConcatTwo-16             98.3ns ± 0%  84.0ns ± 6%  -14.58%  (p=0.008 n=5+5)
ConcatSmall-16            250ns ± 1%   259ns ± 2%   +3.91%  (p=0.008 n=5+5)
ConcatLarge-16            713µs ± 2%   712µs ± 5%     ~     (p=0.421 n=5+5)
```

**Comment by benhoyt** [comment]
Let's go with `Concat2` and `ConcatMulti`. Could you please rename `MultiIndex` to `IndexMulti` while you're at it, to be consistent with that ordering?

**Comment by raff** [comment]
This is the output of benchstat.sh. Not sure why there is a bunch of tests that every time I run it show up as slower than before (and completely unrelated to string concatenation):

```
name                     old time/op  new time/op   delta
ForInLoop-16             1.19µs ± 6%   1.33µs ± 2%  +11.44%  (p=0.008 n=5+5)
BuiltinGsub-16           12.6µs ± 0%   14.0µs ± 7%  +10.87%  (p=0.016 n=4+5)
BuiltinIndex-16           834ns ± 1%    904ns ± 3%   +8.31%  (p=0.008 n=5+5)
ForLoop-16               1.77µs ± 2%   1.92µs ± 7%   +8.24%  (p=0.016 n=5+5)
GlobalVars-16             200ns ± 6%    215ns ± 3%   +7.59%  (p=0.008 n=5+5)
BuiltinSplitRegex-16     68.6µs ± 0%   73.5µs ± 8%   +7.22%  (p=0.008 n=5+5)
LocalVars-16              221ns ± 5%    236ns ± 3%   +6.80%  (p=0.016 n=5+5)
BuiltinMatch-16          2.09µs ± 2%   2.21µs ± 4%   +5.82%  (p=0.016 n=5+5)
BinaryOperators-16        322ns ± 0%    341ns ± 1%   +5.74%  (p=0.008 n=5+5)
BuiltinSubAmpersand-16   10.5µs ± 1%   11.0µs ± 5%   +5.15%  (p=0.008 n=5+5)
BuiltinSubstr-16          568ns ± 1%    597ns ± 6%   +5.08%  (p=0.016 n=5+5)
IfStatement-16            107ns ± 1%    112ns ± 1%   +4.56%  (p=0.008 n=5+5)
CondExpr-16               106ns ± 2%    110ns ± 3%   +4.47%  (p=0.008 n=5+5)
BuiltinSub-16            10.6µs ± 1%   11.0µs ± 4%   +4.05%  (p=0.016 n=5+5)
ArrayOperations-16        434ns ± 1%    442ns ± 1%   +1.77%  (p=0.008 n=5+5)
Comparisons-16            245ns ± 0%    250ns ± 2%   +1.70%  (p=0.008 n=5+5)
AugAssign-16              755ns ± 1%    766ns ± 1%   +1.49%  (p=0.008 n=5+5)
IncrDecr-16               114ns ± 4%    110ns ± 7%     ~     (p=0.222 n=5+5)
SimpleBuiltins-16         595ns ± 6%    564ns ±11%     ~     (p=0.222 n=5+5)
BuiltinLength-16          334ns ± 2%    328ns ± 4%     ~     (p=0.278 n=5+5)
BuiltinSplitSpace-16     28.4µs ± 2%   28.7µs ± 6%     ~     (p=0.730 n=5+5)
BuiltinGsubAmpersand-16  12.8µs ± 2%   13.0µs ± 2%     ~     (p=0.095 n=5+5)
BuiltinSprintf-16        7.20µs ± 2%   7.23µs ± 4%     ~     (p=1.000 n=5+5)
RecursiveFunc-16         11.2µs ± 4%   11.3µs ± 6%     ~     (p=0.548 n=5+5)
FuncCall-16              1.99µs ± 1%   2.01µs ± 3%     ~     (p=0.690 n=5+5)
NativeFunc-16            8.08µs ± 2%   8.40µs ± 7%     ~     (p=0.056 n=5+5)
SimplePattern-16         97.8ns ± 3%  101.5ns ± 5%     ~     (p=0.151 n=5+5)
GetField-16               316ns ± 3%    319ns ± 4%     ~     (p=0.548 n=5+5)
SetField-16               640ns ± 0%    642ns ± 2%     ~     (p=0.730 n=4+5)
RegexMatch-16             856ns ± 0%    874ns ± 2%     ~     (p=0.111 n=4+5)
Assign-16                 202ns ± 2%    200ns ± 1%     ~     (p=0.341 n=5+5)
ConcatSmall-16            517ns ± 0%    263ns ± 4%  -49.15%  (p=0.008 n=5+5)
ConcatLarge-16           5.22ms ± 1%   0.71ms ± 1%  -86.46%  (p=0.008 n=5+5)
[Geo mean]               1.46µs        1.28µs       -12.52%
```

**Comment by benhoyt** [comment]
Yeah, benchmarks are funny sometimes. You have to run them on a quiet machine, and even then there can be fluctuation due to the compiler moving things around, or the x86 architecture being (too) clever. But here's what I'm seeing, and it's only split/sub that have slowed down a little, so I'm comfortable with this:

```
name                    old time/op  new time/op  delta
BuiltinSplitRegex-8     87.7µs ± 0%  94.7µs ± 9%   +7.98%  (p=0.008 n=5+5)
BuiltinSplitSpace-8     35.1µs ± 0%  37.6µs ± 0%   +7.10%  (p=0.008 n=5+5)
BuiltinGsubAmpersand-8  16.2µs ± 0%  17.3µs ± 6%   +6.56%  (p=0.008 n=5+5)
BuiltinSubAmpersand-8   13.6µs ± 0%  14.4µs ± 2%   +5.85%  (p=0.008 n=5+5)
BuiltinSub-8            13.6µs ± 0%  14.3µs ± 3%   +5.45%  (p=0.008 n=5+5)
BuiltinGsub-8           16.2µs ± 0%  16.6µs ± 0%   +2.38%  (p=0.016 n=5+4)
SetField-8               810ns ± 0%   829ns ± 1%   +2.33%  (p=0.008 n=5+5)
BuiltinMatch-8          2.71µs ± 1%  2.69µs ± 3%     ~     (p=0.635 n=5+5)
BuiltinSprintf-8        9.23µs ± 0%  9.27µs ± 1%     ~     (p=0.667 n=5+5)
RecursiveFunc-8         14.9µs ± 0%  14.7µs ± 1%   -1.21%  (p=0.008 n=5+5)
BinaryOperators-8        413ns ± 0%   407ns ± 1%   -1.67%  (p=0.008 n=5+5)
NativeFunc-8            10.8µs ± 0%  10.5µs ± 0%   -2.44%  (p=0.008 n=5+5)
FuncCall-8              2.76µs ± 2%  2.66µs ± 1%   -3.46%  (p=0.008 n=5+5)
CondExpr-8               151ns ± 0%   144ns ± 0%   -4.35%  (p=0.008 n=5+5)
GetField-8               435ns ± 2%   416ns ± 1%   -4.40%  (p=0.008 n=5+5)
IfStatement-8            152ns ± 0%   143ns ± 0%   -5.53%  (p=0.008 n=5+5)
SimpleBuiltins-8         746ns ± 0%   703ns ± 1%   -5.87%  (p=0.008 n=5+5)
RegexMatch-8            1.11µs ± 4%  1.05µs ± 3%   -5.89%  (p=0.032 n=5+5)
BuiltinIndex-8          1.11µs ± 1%  1.04µs ± 0%   -6.05%  (p=0.008 n=5+5)
AugAssign-8              982ns ± 0%   920ns ± 0%   -6.32%  (p=0.008 n=5+5)
ArrayOperations-8        565ns ± 0%   528ns ± 1%   -6.67%  (p=0.008 n=5+5)
SimplePattern-8          134ns ± 0%   125ns ± 1%   -6.69%  (p=0.008 n=5+5)
BuiltinLength-8          429ns ± 0%   400ns ± 2%   -6.70%  (p=0.008 n=5+5)
ForLoop-8               2.50µs ± 0%  2.33µs ± 3%   -6.78%  (p=0.008 n=5+5)
LocalVars-8              300ns ± 2%   279ns ± 1%   -6.93%  (p=0.008 n=5+5)
Comparisons-8            321ns ± 0%   298ns ± 0%   -7.27%  (p=0.029 n=4+4)
GlobalVars-8             269ns ± 1%   249ns ± 3%   -7.58%  (p=0.008 n=5+5)
Assign-8                 261ns ± 0%   240ns ± 1%   -7.76%  (p=0.008 n=5+5)
BuiltinSubstr-8          765ns ± 0%   705ns ± 1%   -7.92%  (p=0.008 n=5+5)
ForInLoop-8             1.78µs ± 4%  1.63µs ± 3%   -8.08%  (p=0.008 n=5+5)
IncrDecr-8               148ns ± 0%   136ns ± 1%   -8.22%  (p=0.016 n=4+5)
ConcatSmall-8            691ns ± 1%   341ns ± 3%  -50.72%  (p=0.008 n=5+5)
ConcatLarge-8           8.35ms ± 2%  1.10ms ± 3%  -86.88%  (p=0.008 n=5+5)
[Geo mean]              1.94µs       1.60µs       -17.75%
```

**Comment by benhoyt** [comment]
Thanks again for the contribution, much appreciated! I'll merge this shortly. Actually, looking at this diff now I think I'll revert `Concat2` back to just `Concat` like all the other binary ops (I'll do this after the merge, so don't worry).

**Comment by benhoyt** [comment]
Final tweaks and updating benchmark results (from the run on my machine): https://github.com/benhoyt/goawk/commit/bd053c958b3dd1a4460768d1b344ee610059edd6

**Pull request merged**